marlowe-playground-client/src/Marlowe/Parser.purs

```diff
@@ -49,17 +73,58 @@ instance prettyTerm :: Pretty a => Pretty (Term a) where
   prettyFragment (Term a) = prettyFragment a
   prettyFragment (Hole name _ _ _) = Leijen.text $ "?" <> name
 
-hole :: forall a. Parser String (Term a)
-hole = do
-  (ParseState _ start _) <- MonadState.get
-  name <- fromCharArray <$> (char '?' *> many alphaNum)
-  (ParseState _ end _) <- MonadState.get
-  pure $ Hole name Proxy start end
+-- a concrete type for holes only
+data MarloweHole = MarloweHole String MarloweType Position Position
```

Comment:
Given you've got `Position` twice, I think there's a lot to be said for making this a record with `start :: Position` and `end :: Position`.